### DISCUSSION TITLE
The same events are added every time the script runs

### DISCUSSION THREAD

**Discussion by jaspersj** [derekantrican/GAS-ICS-Sync#402] [General]
Wondering if I've done something wrong that is resulting in every calendar event being added to the calendar repeatedly every time the script runs.  For example, one event is on the calendar for the same day 225 times.  This is the same for all events from sync'd calendar.  Thanks.

**Comment by jonas0b1011001** [comment]
Can you share one of the script's execution logs from [https://script.google.com/home/executions](https://script.google.com/home/executions)?

**Reply by jaspersj** [reply]
Thanks for the reply.  When I go to the execution list, the cloud log selection is greyed out for each of the individual runs.  Is there another way to dump the execution? When I just expand the execution themselves, I've noticed that the repeating pattern is that one execution times out after ~360 seconds and then the next (set to refresh every 5 minutes) shows as Completed but the log indicates that it didn't because the previous iteration was still running.  I'm not sure why it would take ~360 seconds to run in the first place.  I'll paste in my settings below.  Thanks!!!

<img width="1198" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/f37f3f3a-a424-4927-a0a3-107b53ec60fe">

```
var howFrequent = 5;                     // What interval (minutes) to run this script on to check for new events.  Any integer can be used, but will be rounded up to 5, 10, 15, 30 or to the nearest hour after that.. 60, 120, etc. 1440 (24 hours) is the maximum value.  Anything above that will be replaced with 1440.
var onlyFutureEvents = false;             // If you turn this to "true", past events will not be synced (this will also removed past events from the target calendar if removeEventsFromCalendar is true)
var addEventsToCalendar = true;           // If you turn this to "false", you can check the log (View > Logs) to make sure your events are being read correctly before turning this on
var modifyExistingEvents = true;          // If you turn this to "false", any event in the feed that was modified after being added to the calendar will not update
var removeEventsFromCalendar = true;      // If you turn this to "true", any event created by the script that is not found in the feed will be removed.
var removePastEventsFromCalendar = true;  // If you turn this to "false", any event that is in the past will not be removed.
var addAlerts = "yes";                    // Whether to add the ics/ical alerts as notifications on the Google Calendar events or revert to the calendar's default reminders ("yes", "no", "default").
var addOrganizerToTitle = false;          // Whether to prefix the event name with the event organiser for further clarity
var descriptionAsTitles = false;          // Whether to use the ics/ical descriptions as titles (true) or to use the normal titles as titles (false)
var addCalToTitle = false;                // Whether to add the source calendar to title
var addAttendees = false;                 // Whether to add the attendee list. If true, duplicate events will be automatically added to the attendees' calendar.
var defaultAllDayReminder = -1;           // Default reminder for all day events in minutes before the day of the event (-1 = no reminder, the value has to be between 0 and 40320)
                                          // See https://github.com/derekantrican/GAS-ICS-Sync/issues/75 for why this is neccessary.
var overrideVisibility = "";              // Changes the visibility of the event ("default", "public", "private", "confidential"). Anything else will revert to the class value of the ICAL event.
var addTasks = false;

var emailSummary = false;                 // Will email you when an event is added/modified/removed to your calendar
var email = "";                           // OPTIONAL: If "emailSummary" is set to true or you want to receive update notifications, you will need to provide your email address
var customEmailSubject = "";              // OPTIONAL: If you want to change the email subject, provide a custom one here. Default: "GAS-ICS-Sync Execution Summary"
var dateFormat = "YYYY-MM-DD"             // date format in the email summary (e.g. "YYYY-MM-DD", "DD.MM.YYYY", "MM/DD/YYYY". separators are ".", "-" and "/")
```

**Reply by jonas0b1011001** [reply]
> Is there another way to dump the execution? When I just expand the execution themselves, [...]

That's what i was looking for.

The script does not even finish getting the existing events from the target calendar before it times out.
Can you please sync to an empty target calendar and see if that completes successfully.

**Reply by jaspersj** [reply]
ok, so updated settings to a new calendar and also updated interval to 15 min vs 5 min.  It ran for ~72 seconds, then failed.  The calendar was created and 114 events were added, but then it threw a DNS error ( screen shots minus each of the individual event lines is below).  I did go back and find that it was running similarly after initially installing the script, although much longer because it removes duplicates after adding them (i'll add a snipit of that below:

<img width="1415" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/428bb40b-66d1-41bb-aebc-3be1e4f76c9d">

<img width="654" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/d67d0894-4e6f-4446-96f8-f51af6cf36dd">

The last time it didn't time out last week:

<img width="1212" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/79a8f86c-7001-4f3d-9bbf-f184c31420f6">

<img width="705" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/56f333f4-df83-4b7f-8a4a-c2fa45d896be">

Thanks for your assistance!!!

**Reply by jonas0b1011001** [reply]
You have to remove the unused sample entries in `var sourceCalendars` aswell as the `,` at the end of your remaining entry.
https://github.com/derekantrican/GAS-ICS-Sync/blob/17cde1a768ad4c6d5ede0e7401e20606fff0be2a/Code.gs#L29-L30

it should look like this:
```
var sourceCalendars = [
  ["https://xyz.ics", "targetCalendar"]
];
```

**Reply by jaspersj** [reply]
That worked!!! Thanks so much for the help!!

**Reply by jaspersj** [reply]
hi @jonas0b1011001  - hoping you can help again.  I've noticed that over the last few days, the script execution time has crept back up to right under and now in many cases right over the 360 sec mark with very little change to the number of events being processed.  Any thoughts?

<img width="1185" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/311da69f-b5c0-409c-96ad-518c0269d6e4">

Today:
<img width="1201" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/edd819b1-b7c0-4434-87b4-2d4725dc70dc">

A week ago when it was taking half the time:
<img width="1225" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/8a45f6c4-be23-4fb5-a186-47258a7f6f6c">

Thanks in advance!!

**Reply by jonas0b1011001** [reply]
> A week ago when it was taking half the time: <img alt="image" width="1225" src="https://private-user-images.githubusercontent.com/13245968/296030493-8a45f6c4-be23-4fb5-a186-47258a7f6f6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDUwMDIwMTIsIm5iZiI6MTcwNTAwMTcxMiwicGF0aCI6Ii8xMzI0NTk2OC8yOTYwMzA0OTMtOGE0NWY2YzQtYmUyMy00ZmI1LWExODYtNDcyNThhN2Y2ZjZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTExVDE5MzUxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIwYTkwZDdmYzlkYWE3YTdkNGY5N2U2YTY5Y2UxMGI2YmI2Yjg4OWFlYmI2ZWNiMjIwOGIwZThmMGQyNzYwNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wGngHJvJQaUFC2nAA0y1Vi0p2miytQp448PRgbDRbwQ">

The fact that the script is adding _new_ events, even though the target and source calendar both have 114 events, makes me believe that your source ics is using non persistent event ids.

Please have a look at #248 and use the modified script linked in the comments (you can ignore the `limitSyncRange` setting). Sync your source ics to an empty calendar, subsequent executions should complete within seconds.

**Reply by jaspersj** [reply]
Awesome - worked like a charm.  The first run took just over a minute and the second with no changes took only 5 seconds.  Thanks again for your assistance.

<img width="1031" alt="image" src="https://github.com/derekantrican/GAS-ICS-Sync/assets/13245968/dbb66596-e563-4a35-84e9-b78adb6a317c">
